test/abstract-sql/test.ts

```diff
@@ -0,0 +1,192 @@
+import * as AbstractSQLCompiler from '../..';
+
+import { expect } from 'chai';
+import _ from 'lodash';
+
+type ExpectedBindings = Array<
+	AbstractSQLCompiler.Binding | AbstractSQLCompiler.Binding[]
+>;
+const bindingsTest = function (
+	actualBindings: AbstractSQLCompiler.Binding[],
+	expectedBindings: AbstractSQLCompiler.Binding[] | false = false,
+) {
+	if (expectedBindings === false) {
+		it('should not have any bindings', () => {
+			expect(actualBindings).to.be.empty;
+		});
+	} else {
+		it('should have matching bindings', () => {
+			expect(actualBindings).to.deep.equal(expectedBindings);
+		});
+	}
+};
+
+type SqlEquals = (
+	actual:
+		| AbstractSQLCompiler.SqlResult
+		| AbstractSQLCompiler.SqlResult[]
+		| string,
+	expected: string,
+) => void;
+const equals: SqlEquals = (actual, expected) => {
+	if (typeof actual !== 'string') {
+		if (Array.isArray(actual)) {
+			throw new Error('Expected a single query, got multiple');
+		}
+		actual = actual.query;
+	}
```

Comment:
I took a minor liberty in this section so that it's possible to do `sqlEquals(result, '...')` and not have to worry about the `SqlResult | [SqlResult, SqlResult]` typing in every single place when `[SqlResult, SqlResult]` only applies to `UpsertNode`s (very very uncommon case)

Reply:
I guess this is never triggered though as the tests pass. Why even allow for `SqlResult[]` if we always throw?